[retake2][mobile] Fix lightweight dispatch OOM error by introducing selective build

larryliu0820 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #80791

To fix #78540 I committed #78983 which is reverted due to internal CI failure. Then I comitted #79215 which was only fixing the failure but didn't have the full feature of #78983. This PR is another try.

This PR adds script to dump all operators from test models and automatically write into `lightweight_dispatch_ops.yaml`. This way we don't have to manually update the yaml file.

larryliu0820 committed:
[retake2][mobile] Fix lightweight dispatch OOM error by introducing selective build

To fix #78540 I committed #78983 which is reverted due to internal CI failure. Then I commited #79215 which was only fixing the failure but didn't have the full feature of #78983. This PR is another try.

[ghstack-poisoned]

larryliu0820 committed:
Update on "[retake2][mobile] Fix lightweight dispatch OOM error by introducing selective build"


To fix #78540 I committed #78983 which is reverted due to internal CI failure. Then I commited #79215 which was only fixing the failure but didn't have the full feature of #78983. This PR is another try.

[ghstack-poisoned]

ezyang commented on test/mobile/lightweight_dispatch/lightweight_dispatch_ops.yaml:
```diff
@@ -0,0 +1,60 @@
+- aten::upsample_nearest2d.vec
+- aten::upsample_nearest3d.vec
+- aten::upsample_trilinear3d.vec
+- aten::zeros
```
How do we maintain this list going forward?

larryliu0820 replied:
This is a dump of all the operators used in the test. It would be hard for someone without context to maintain it. I guess next step I can write a script to automatically dump all the operators from the tests. We have similar script already so this should be feasible

larryliu0820 replied:
Added script to dump all operators in test models into yaml file.

ezyang commented:
What is the proximal cause of the OOM? Is there some memory hungry algorithm in the dispatch implementation?

larryliu0820 commented:
> What is the proximal cause of the OOM? Is there some memory hungry algorithm in the dispatch implementation?


My intuition is we are generating more files in static dispatch mode, it just takes more memory to compile/link the whole binary.

larryliu0820 committed:
Update on "[retake2][mobile] Fix lightweight dispatch OOM error by introducing selective build"


To fix #78540 I committed #78983 which is reverted due to internal CI failure. Then I commited #79215 which was only fixing the failure but didn't have the full feature of #78983. This PR is another try.

[ghstack-poisoned]

larryliu0820 committed:
Update on "[retake2][mobile] Fix lightweight dispatch OOM error by introducing selective build"


To fix #78540 I committed #78983 which is reverted due to internal CI failure. Then I comitted #79215 which was only fixing the failure but didn't have the full feature of #78983. This PR is another try.

This PR adds script to dump all operators from test models and automatically write into `lightweight_dispatch_ops.yaml`. This way we don't have to manually update the yaml file.

[ghstack-poisoned]

larryliu0820 committed:
Update on "[retake2][mobile] Fix lightweight dispatch OOM error by introducing selective build"


To fix #78540 I committed #78983 which is reverted due to internal CI failure. Then I comitted #79215 which was only fixing the failure but didn't have the full feature of #78983. This PR is another try.

This PR adds script to dump all operators from test models and automatically write into `lightweight_dispatch_ops.yaml`. This way we don't have to manually update the yaml file.

[ghstack-poisoned]

larryliu0820 committed:
Update on "[retake2][mobile] Fix lightweight dispatch OOM error by introducing selective build"


To fix #78540 I committed #78983 which is reverted due to internal CI failure. Then I comitted #79215 which was only fixing the failure but didn't have the full feature of #78983. This PR is another try.

This PR adds script to dump all operators from test models and automatically write into `lightweight_dispatch_ops.yaml`. This way we don't have to manually update the yaml file.

[ghstack-poisoned]

larryliu0820 committed:
Update on "[retake2][mobile] Fix lightweight dispatch OOM error by introducing selective build"


To fix #78540 I committed #78983 which is reverted due to internal CI failure. Then I comitted #79215 which was only fixing the failure but didn't have the full feature of #78983. This PR is another try.

This PR adds script to dump all operators from test models and automatically write into `lightweight_dispatch_ops.yaml`. This way we don't have to manually update the yaml file.

[ghstack-poisoned]

raziel commented on test/mobile/lightweight_dispatch/tests_setup.py:
```diff
@@ -1,203 +1,143 @@
+            ModelWithMultipleOps(),
+        ]
+        shutil.copyfile(ops_yaml, backup)
+        with open(ops_yaml, 'a') as f:
```
don't you need to prepend a \n? or hardcode that in the .yaml file?

larryliu0820 replied:
[lightweight_dispatch_ops.yaml](https://github.com/pytorch/pytorch/pull/80791/files#diff-2e57f80e10ce026e5af291cd7f4493a621a6ba5017b0ae619ebfa4dbfc763fbf) actually has a trailing newline.

raziel reviewed (approved):
LGTM
Just check my comment about the new line.

larryliu0820 commented:
@pytorchbot merge
